src/librustc_lint/unused.rs

```diff
@@ -270,47 +270,52 @@ impl UnusedParens {
                                 msg: &str,
                                 struct_lit_needs_parens: bool) {
```

Comment:
Nit: this name no longer seems very accurate; let's call it something like `is_restricted_expr: bool` perhaps? And I think a comment on `check_unused_parens_core` would be good. Something like:

---

Invoked on expressions that appear in places where people traditionally put parentheses out of habit, but where they are not needed in Rust. For example, `if (a == b) { .. }`. Issues a lint if `value` is a parenthesized node, unless those parentheses are deemed "necessary".

Parentheses are considered necessary if (a) `is_restricted_expr` is true, which indicates that this is a context (such as an `if`) where only a subset of expressions are permitted and (b) the parenthesized expression falls outside of that subset. So e.g. `if (Foo { a: 22 }) { .. }` would not get a warning; similarly `while (break) { .. }` would not.